File: src/main/java/com/ndgndg91/ordermatchingengine/order/OrderEntry.java
package com.ndgndg91.ordermatchingengine.order;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.ToString;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Getter
@ToString
public class OrderEntry {
    private final String orderId;
    private final OrderType orderType;
    private final int shares;
    private final PriceType priceType;
    private final BigDecimal price;
    private final LocalDateTime timestamp;
    private boolean partialMatched;
    private final List<PartialMatched> partialMatchedEntries = new ArrayList<>();

    public OrderEntry(final Order order) {
        this.orderId = order.getOrderId();
        this.orderType = order.getOrderType();
        this.shares = order.getShares();
        this.priceType = order.getPriceType();
        this.price = order.getPrice();
        this.timestamp = order.getTimestamp();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        OrderEntry that = (OrderEntry) o;
        return Objects.equals(orderId, that.orderId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(orderId);
    }

    public void partialMatched(final OrderEntry e) {
        this.partialMatched = true;
        PartialMatched pm = new PartialMatched(
                e.orderId,
                e.orderType,
                e.shares(),
                e.priceType,
                e.price,
                e.timestamp
        );
        this.partialMatchedEntries.add(pm);
    }

    public void partialMatched(final OrderEntry e, final int shares) {
        this.partialMatched = true;
        PartialMatched pm = new PartialMatched(
                e.orderId,
                e.orderType,
                shares,
                e.priceType,
                e.price,
                e.timestamp
        );
        this.partialMatchedEntries.add(pm);
    }

    public int shares() {
        if (partialMatched) {
            return shares - partialMatchedEntries.stream().parallel().mapToInt(p -> p.shares).sum();
        } else {
            return shares;
        }
    }

    public int partialShares() {
        return partialMatchedEntries.stream().parallel().mapToInt(p -> p.shares).sum();
    }

    @RequiredArgsConstructor
    @ToString
    private static final class PartialMatched {
        private final String orderId;
        private final OrderType orderType;
        private final int shares;
        private final PriceType priceType;
        private final BigDecimal price;
        private final LocalDateTime timestamp;
    }
}

File: OrderMatchingEngineKotlin/src/main/kotlin/com/ndgndg91/ordermatchingenginekotlin/order/OrderEntry.kt
package com.ndgndg91.ordermatchingenginekotlin.order

import java.math.BigDecimal
import java.time.LocalDateTime

open class OrderEntry(
    open val orderId: String,
    open val orderType: OrderType,
    open val shares: Int,
    open val priceType: PriceType,
    open val price: BigDecimal,
    open val timestamp: LocalDateTime,
    open var partialMatched: Boolean,
    open var partialMatchedEntries: List<PartialMatched>) {

    fun toBidEntry(): BidOrderEntry = BidOrderEntry(
        orderId,
        orderType,
        shares,
        priceType,
        price,
        timestamp,
        partialMatched,
        partialMatchedEntries
    )

    fun toAskEntry(): AskOrderEntry = AskOrderEntry(
        orderId,
        orderType,
        shares,
        priceType,
        price,
        timestamp,
        partialMatched,
        partialMatchedEntries
    )


    fun partialMatched(e: OrderEntry) {
        this.partialMatched = true
        val pm = PartialMatched(
            e.orderId,
            e.orderType,
            e.shares(),
            e.priceType,
            e.price,
            e.timestamp
        )
        this.partialMatchedEntries = this.partialMatchedEntries.plus(pm)
    }

    fun partialMatched(e: OrderEntry, shares: Int) {
        this.partialMatched = true
        val pm = PartialMatched(e.orderId, e.orderType, shares, e.priceType, e.price, e.timestamp)
        this.partialMatchedEntries = this.partialMatchedEntries.plus(pm)
    }

    fun shares(): Int {
        return if (partialMatched) {
            this.shares - partialMatchedEntries.asSequence().map { it.shares }.sum()
        } else {
            this.shares
        }
    }

    fun partialShares(orderId: String): Int {
        val sum = this.partialMatchedEntries.asSequence()
            .filter { it.orderId == orderId }
            .map { it.shares }
            .sum()

        return if (sum > 0) sum else this.shares - partialMatchedEntries.asSequence()
            .map { it.shares }
            .sum()
    }


    data class PartialMatched(
        val orderId: String,
        val orderType: OrderType,
        val shares: Int,
        val priceType: PriceType,
        val price: BigDecimal,
        val timestamp: LocalDateTime
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as OrderEntry

        if (orderId != other.orderId) return false

        return true
    }

    override fun hashCode(): Int {
        return orderId.hashCode()
    }

    override fun toString(): String {
        return "OrderEntry(orderId='$orderId', orderType=$orderType, shares=$shares, priceType=$priceType, price=$price, timestamp=$timestamp, partialMatched=$partialMatched, partialMatchedEntries=$partialMatchedEntries)"
    }
}